graph/vertex/impl/AttributeVertexImpl.java

```diff
@@ -27,282 +27,522 @@
 import com.vaticle.typedb.core.graph.iid.IndexIID;
 import com.vaticle.typedb.core.graph.iid.VertexIID;
 import com.vaticle.typedb.core.graph.vertex.AttributeVertex;
+import com.vaticle.typedb.core.graph.vertex.ThingVertex;
 
 import java.time.LocalDateTime;
 
 import static com.vaticle.typedb.common.util.Objects.className;
 import static com.vaticle.typedb.core.common.exception.ErrorMessage.ThingRead.INVALID_THING_VERTEX_CASTING;
 import static com.vaticle.typedb.core.common.exception.ErrorMessage.Transaction.ILLEGAL_OPERATION;
 
-public abstract class AttributeVertexImpl<VALUE> extends ThingVertexImpl implements AttributeVertex<VALUE> {
+public abstract class AttributeVertexImpl {
 
-    private final VertexIID.Attribute<VALUE> attributeIID;
-    private java.lang.Boolean isPersisted;
+    public static abstract class Read<VALUE> extends ThingVertexImpl.Read implements AttributeVertex.Read<VALUE> {
 
-    AttributeVertexImpl(ThingGraph graph, VertexIID.Attribute<VALUE> iid, boolean isInferred) {
-        super(graph, iid, isInferred);
-        this.attributeIID = iid;
-        this.isPersisted = null;
-    }
+        private final VertexIID.Attribute<VALUE> attributeIID;
+        private java.lang.Boolean isPersisted;
+
+        Read(ThingGraph graph, VertexIID.Attribute<VALUE> iid) {
+            super(graph, iid);
+            this.attributeIID = iid;
+            this.isPersisted = null;
+        }
+
+        public static AttributeVertexImpl.Read<?> of(ThingGraph graph, VertexIID.Attribute<?> iid) {
+            switch (iid.valueType()) {
+                case BOOLEAN:
+                    return new AttributeVertexImpl.Read.Boolean(graph, iid.asBoolean());
+                case LONG:
+                    return new AttributeVertexImpl.Read.Long(graph, iid.asLong());
+                case DOUBLE:
+                    return new AttributeVertexImpl.Read.Double(graph, iid.asDouble());
+                case STRING:
+                    return new AttributeVertexImpl.Read.String(graph, iid.asString());
+                case DATETIME:
+                    return new AttributeVertexImpl.Read.DateTime(graph, iid.asDateTime());
+                default:
+                    assert false;
+                    return null;
+            }
+        }
+
+        @Override
+        public VertexIID.Attribute<VALUE> iid() {
```

Comment:
from here downward is a fair amount of duplication, but also omits some things that are in the `AttributeVertexImpl.Write` (eg. `isInferred`)